test/unit/org/apache/cassandra/cql3/validation/entities/TupleTypeTest.java

```diff
@@ -233,5 +256,149 @@ public void testReversedTypeTuple() throws Throwable
         DateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mmX", Locale.ENGLISH);
         assertRows(execute("SELECT tdemo FROM %s"), row(tuple( df.parse("2017-02-03 03:05+0000"), "Europe")));
     }
+
+    @Test
+    public void tuplePkReadWrite()
+    {
+        int numRows = 10;
+        class Case {
+            TupleType type;
+            List<ByteBuffer> uniqueRows;
+        }
+        Gen<TupleType> baseGen = getTupleType();
+        Gen<Case> gen = rnd -> {
+            Case c = new Case();
+            c.type = baseGen.generate(rnd);
+            TypeSupport<ByteBuffer> support = getTypeSupport(c.type);
+            Gen<ByteBuffer> valueGen = Generators.filter(support.valueGen, b -> b.remaining() <= Short.MAX_VALUE);
+            Set<ByteBuffer> distinctRows = new HashSet<>(numRows);
+            boolean done = false;
+            for (int i = 0; !done && i < numRows; i++)
```

Comment:
nit - could just write as a simpler loop - you bound the total number of generation attempts to get the list you want that way. I don't think `HashSet.size()` is expensive
```
final int OVER_GEN_FACTOR = 10;
for (int i = 0; distinctRows.size() < numRows && i < OVER_GEN_FACTOR * numRows; i++)
{
    distinctRows.add(valueGen.generate(rnd)
}
if (distinctRows.size() != numRows)
{
    System.err.println("Not enough uniq row for type " + c.type + ", returning with " + i + "rows");
}